ui/src/utils/auto_updater.rs

```diff
@@ -0,0 +1,206 @@
+pub struct GitHubRelease {
+    pub tag_name: String,
+    assets: Vec<GitHubAsset>,
+}
+
+#[derive(Debug, Deserialize, Clone)]
+struct GitHubAsset {
+    name: String,
+    browser_download_url: String,
+    size: usize,
+}
+
+pub fn get_download_dest() -> Option<PathBuf> {
+    match FileDialog::new()
+        .set_directory(dirs::home_dir().unwrap_or(".".into()))
+        .set_title(&get_local_text("uplink.pick-download-directory"))
+        .pick_folder()
+    {
+        Some(x) => Some(x),
+        None => {
+            log::debug!("update download canceled by user");
+            None
+        }
+    }
+}
+
+pub async fn check_for_release() -> anyhow::Result<Option<GitHubRelease>> {
+    let latest_release =
+        get_github_release("https://api.github.com/repos/Satellite-im/Uplink/releases/latest")
+            .await?;
+
+    if versions_match(&latest_release.tag_name) {
+        Ok(None)
+    } else {
+        Ok(Some(latest_release))
+    }
+}
+
+pub async fn download_update(
+    binary_dest: PathBuf,
+    ch: mpsc::UnboundedSender<f32>,
+) -> anyhow::Result<String> {
+    let latest_release =
+        get_github_release("https://api.github.com/repos/Satellite-im/Uplink/releases/latest")
+            .await?;
+    let find_asset = |name: &str| {
+        latest_release
+            .assets
+            .iter()
+            .find(|x| x.name.contains(name))
+            .cloned()
+            .ok_or(anyhow::format_err!("failed to find {name}"))
+    };
+
+    let binary_asset = if cfg!(target_os = "windows") {
+        find_asset(".msi")?
+    } else if cfg!(target_os = "linux") {
+        find_asset(".deb")?
+    } else if cfg!(target_os = "macos") {
+        find_asset("Uplink-Mac-Universal.zip")?
+    } else {
+        bail!("unknown OS type. failed to find binary");
+    };
+
+    let total_download_size = binary_asset.size as f32;
+    let client = get_client()?;
+    let (tx, mut rx) = mpsc::unbounded_channel::<anyhow::Result<usize>>();
+    tokio::spawn(async move {
+        if let Err(e) = download_file(
+            &client,
+            binary_dest.join(&binary_asset.name),
+            &binary_asset.browser_download_url,
+            tx.clone(),
+        )
+        .await
+        {
+            let _ = tx.send(Err(e));
+        }
+    });
+
+    let mut total_bytes_downloaded = 0.0_f32;
+    while let Some(x) = rx.recv().await {
+        match x {
+            Ok(b) => {
+                total_bytes_downloaded += b as f32;
+                let _ = ch.send(100_f32 * total_bytes_downloaded / total_download_size);
+            }
+            Err(e) => return Err(e),
+        }
+    }
+
+    Ok(latest_release.tag_name)
+}
+
+fn get_client() -> Result<Client, reqwest::Error> {
+    let mut headers = header::HeaderMap::new();
+    headers.insert(
+        "User-Agent",
+        header::HeaderValue::from_static(
+            "Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/111.0",
```

Comment:
I'm willing to do that, but I worry that the GitHub api wants to only be accessed by a browser. When I didn't set any user agent it rejected my request. I can test it out and see what happens.

Reply:
The last I checked, their API required a user agent otherwise it would get rejected. The agent name can be anything (such as the name of the project).

Reply:
https://docs.github.com/en/rest/overview/resources-in-the-rest-api?apiVersion=2022-11-28#user-agent-required